File: tinDB/src/main/DatabaseSpecification/Database/DatabaseEdge.java
package Database;

// class specifying a database edge.
// this edge is now labeled (might be changed later or depending on the database -> use a bool directedGraph (?))
// this edge is a 3-tuple with source and target node and the label of the edge.
public class DatabaseEdge {

    public DatabaseNode source;
    public DatabaseNode target;
    public String label;

    DatabaseEdge(DatabaseNode source, DatabaseNode target, String label) {
        this.source = source;
        this.target = target;
        this.label = label.toLowerCase();
    }

    public String toString() {
        return String.format("(%s) -[%s]-> (%s)", source.identifier, label, target.identifier);
    }

    public void print() {
        System.out.println(toString());
    }
}

File: src/main/kotlin/tin/model/database/DatabaseEdge.kt
package tin.model.database

import tin.model.graph.Edge
import tin.model.query.QueryEdge
import tin.model.query.QueryNode

class DatabaseEdge(
        override var source: DatabaseNode,
        override var target: DatabaseNode,
        label: String
) : Edge(
        source, target, label
) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is DatabaseEdge) return false

        return this.checkForNodesEquality(other) &&
                label == other.label
    }

    /**
     * call the trimmed QueryNode.equals() method in order to prevent a circular dependency.
     */
    private fun checkForNodesEquality(other: DatabaseEdge): Boolean {
        return source.equalsWithoutEdges(other.source) &&
                target.equalsWithoutEdges(other.target)
    }

    override fun hashCode(): Int {
        var result = source.identifier.hashCode()
        result = 31 * result + target.identifier.hashCode()
        result = 31 * result + label.hashCode()
        return result
    }

    override fun toString(): String {
        return String.format("(%s) -[%s]-> (%s)", source.identifier, label, target.identifier)
    }
}